util/linuxfw/nftables_runner_test.go

```diff
@@ -0,0 +1,790 @@
+	if err != nil {
+		t.Fatalf("conn.ListTables() failed: %v", err)
+	}
+
+	if len(tables) != 0 {
+		t.Fatalf("len(tables) = %d, want 0", len(tables))
+	}
+}
+
+func getTsChains(
+	conn *nftables.Conn,
+	proto nftables.TableFamily) (*nftables.Chain, *nftables.Chain, *nftables.Chain, error) {
+	chains, err := conn.ListChainsOfTableFamily(nftables.TableFamilyIPv4)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("list chains failed: %w", err)
+	}
+	var chainInput, chainForward, chainPostrouting *nftables.Chain
+	for _, chain := range chains {
+		switch chain.Name {
+		case "ts-input":
+			chainInput = chain
+		case "ts-forward":
+			chainForward = chain
+		case "ts-postrouting":
+			chainPostrouting = chain
+		}
+	}
+	return chainInput, chainForward, chainPostrouting, nil
+}
+
+// findV4BaseRules verifies that the base rules are present in the input and forward chains.
+func findV4BaseRules(
+	conn *nftables.Conn,
+	inpChain *nftables.Chain,
+	forwChain *nftables.Chain,
+	tunname string) ([]*nftables.Rule, error) {
+	want := []*nftables.Rule{}
+	rule, err := createRangeRule(inpChain.Table, inpChain, tunname, tsaddr.ChromeOSVMRange(), expr.VerdictReturn)
+	if err != nil {
+		return nil, fmt.Errorf("create rule: %w", err)
+	}
+	want = append(want, rule)
+	rule, err = createRangeRule(inpChain.Table, inpChain, tunname, tsaddr.CGNATRange(), expr.VerdictDrop)
+	if err != nil {
+		return nil, fmt.Errorf("create rule: %w", err)
+	}
+	want = append(want, rule)
+	rule, err = createDropOutgoingPacketFromCGNATRangeRuleWithTunname(forwChain.Table, forwChain, tunname)
+	if err != nil {
+		return nil, fmt.Errorf("create rule: %w", err)
+	}
+	want = append(want, rule)
+
+	get := []*nftables.Rule{}
+	for _, rule := range want {
+		getRule, err := findRule(conn, rule)
+		if err != nil {
+			return nil, fmt.Errorf("find rule: %w", err)
+		}
+		get = append(get, getRule)
+	}
+	return get, nil
+}
+
+func findCommonBaseRules(
+	conn *nftables.Conn,
+	forwChain *nftables.Chain,
+	tunname string) ([]*nftables.Rule, error) {
+	want := []*nftables.Rule{}
+	rule, err := createSetSubnetRouteMarkRule(forwChain.Table, forwChain, tunname)
+	if err != nil {
+		return nil, fmt.Errorf("create rule: %w", err)
+	}
+	want = append(want, rule)
+	rule, err = createMatchSubnetRouteMarkRule(forwChain.Table, forwChain, Accept)
+	if err != nil {
+		return nil, fmt.Errorf("create rule: %w", err)
+	}
+	want = append(want, rule)
+	rule = createAcceptOutgoingPacketRule(forwChain.Table, forwChain, tunname)
+	want = append(want, rule)
+
+	get := []*nftables.Rule{}
+	for _, rule := range want {
+		getRule, err := findRule(conn, rule)
+		if err != nil {
+			return nil, fmt.Errorf("find rule: %w", err)
+		}
+		get = append(get, getRule)
+	}
+
+	return get, nil
+}
+
+func TestNFTAddAndDelNetfilterBase(t *testing.T) {
+	if os.Geteuid() != 0 {
+		t.Skip(t.Name(), " requires privileges to create a namespace in order to run")
+		return
+	}
+
+	conn, ns := newSysConn(t)
+	defer cleanupSysConn(t, ns)
```

Comment:
consider having `newSysConn` call `t.Cleanup(cleanupSysConn)` (which will require reworking how `t` and `ns` are stored a little, maybe the cleanup would need to be a closure) to make it a bit easier for test writers

Reply:
Updated